Made before and after "html-processing" events use callback value

ascariandrea opened the pull request:
Hi @jantimon,
I came up with this simple solution to satisfy both needs.
This approach:
 - keep backwards compatibility
 - offer a new api to deal with `html-webpack-plugin-(before|after)-html-processing`

jantimon commented on index.js:
```diff
@@ -138,8 +138,8 @@ HtmlWebpackPlugin.prototype.apply = function (compiler) {
-          .then(function () {
-            return pluginArgs.html;
+          .then(function (newPluginArgs) {
+            return newPluginArgs && newPluginArgs.html || pluginArgs.html;
```
Shouldn't this be `newPluginArgs ? newPluginArgs.html : pluginArgs.html`?
Does this allow to be changed by more than one plugin?
Can you please add a unit test for this?

ascariandrea replied:
> Shouldn't this be newPluginArgs ? newPluginArgs.html : pluginArgs.html

It's the same, or am I missing something?

> Can you please add a unit test for this?

Sure! And I'm going to test if it can be changed by more than one plugin

jantimon replied:
No if you would explicitly return an empty string it would fallback to the pluginArgs (strange use case but still a unexpected behaviour)

jantimon commented on index.js:
```diff
@@ -157,8 +157,8 @@ HtmlWebpackPlugin.prototype.apply = function (compiler) {
-          .then(function () {
-            return pluginArgs.html;
+          .then(function (newPluginArgs) {
+            return newPluginArgs && newPluginArgs.html || pluginArgs.html;
```
Shouldn't this be `newPluginArgs ? newPluginArgs.html : pluginArgs.html`?
Does this allow to be changed by more than one plugin?
Can you please add a unit test for this?

ascariandrea commented:
You’re right!

-- 
Andrea Ascari
dev @ buildo

ph	3405348124
ma	dev.ascariandrea@gmail.com

gh	github.com/ascariandrea
in	linkedin.com/ascariandrea

pi	02568700351

Sent with Airmail

On 26 September 2016 at 13:48:44, Jan Nicklas (notifications@github.com) wrote:

@jantimon commented on this pull request.

In index.js:

> @@ -138,8 +138,8 @@ HtmlWebpackPlugin.prototype.apply = function (compiler) {
       .then(function (html) {
         var pluginArgs = {html: html, assets: assets, plugin: self, outputName: self.childCompilationOutputName};
         return applyPluginsAsyncWaterfall('html-webpack-plugin-before-html-processing', pluginArgs)
-          .then(function () {
-            return pluginArgs.html;
+          .then(function (newPluginArgs) {
+            return newPluginArgs && newPluginArgs.html || pluginArgs.html;
No if you would explicitly return an empty string it would fallback to the pluginArgs (strange use case but still a unexpected behaviour)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

ascariandrea commented:
I turned out all the declared events mutate the object the plugin uses to generate html.
Are you agree in refactor all those methods?

jantimon commented:
Maybe we could move it out to a helper which provides the promified `applyPluginsAsyncWaterfall` helper:

https://github.com/ampedandwired/html-webpack-plugin/blob/9757d3df30653fe52dcc06766322300ca2649615/index.js#L66

ascariandrea commented:
I added an helper to merge passed `pluginArgs` object with the object returned from promise, refactor the logic with this new helper and added tests.

I kept old commits just for reference, but I'm going to delete them if you approve the new changes :)

jantimon commented on index.js:
```diff
@@ -137,28 +137,37 @@ HtmlWebpackPlugin.prototype.apply = function (compiler) {
+          .then(function (result) {
+            var html = result.html;
+            var assets = result.assets;
+            var body = result.body;
```
why do we need this line?

jantimon replied:
Yes would be way better 👍

jantimon commented on index.js:
```diff
@@ -137,28 +137,37 @@ HtmlWebpackPlugin.prototype.apply = function (compiler) {
         // Prepare script and link tags
         var assetTags = self.generateAssetTags(assets);
-        var pluginArgs = {head: assetTags.head, body: assetTags.body, plugin: self, chunks: chunks, outputName: self.childCompilationOutputName};
+        var pluginArgs = {head: assetTags.head, body: assetTags.body, plugin: self, chunks: chunks, outputName: self.childCompilationOutputName, assets: assets, html: html};
```
Why did you add html here?
We don't want people to modify assets or html at this points anymore - this would cause a mess.

ascariandrea replied:
You're right 👍

jantimon commented on spec/BasicSpec.js:
```diff
@@ -870,6 +870,52 @@ describe('HtmlWebpackPlugin', function () {
     });
   });
 
+  it('allows to modify sequentially the html during html-webpack-plugin-after-html-processing event', function (done) {
```
This test is just perfect - can we copy it for your usecase where you don't want to modify the root object?
And one where we mix both?

ascariandrea replied:
Done!

jantimon commented:
Just added a test and it fails:

https://github.com/ampedandwired/html-webpack-plugin/tree/ascariandrea-return-updated-html-on-after-html-processing

ascariandrea commented:
Yep, but it has to fail.
Cause in the docs says:
> Note that the callback must be passed the htmlPluginData in order to pass this onto any other plugins listening on the same 'html-webpack-plugin-before-html-processing' event.

But if you want the this changes to be breaking we can make the returned value mandatory.

jantimon commented:
Otherwise it would not allow two different modules to work together right?

ascariandrea commented:
Yes

https://github.com/webpack/tapable#applypluginsasyncwaterfall

ascariandrea commented:
@jantimon any updates?

jantimon commented:
Still not sure what todo - maybe we could deprecate the usage without a return value first.

ascariandrea commented:
The current behaviour is not breaking, so it could be merged without problem.
Then you can schedule the breaking change by removing the backward compatibility.

jantimon commented on spec/BasicSpec.js:
```diff
@@ -15,6 +15,7 @@ var path = require('path');
 var fs = require('fs');
 var webpack = require('webpack');
 var rimraf = require('rimraf');
+var extend = require('lodash/extend');
```
Can you please use `_.extend` like you used it in the index.js file?

jantimon commented on index.js:
```diff
@@ -612,4 +617,18 @@ HtmlWebpackPlugin.prototype.getAssetFiles = function (assets) {
+  return function (eventName, pluginArgs) {
+    return promisedApplyPluginsAsyncWaterfall(eventName, pluginArgs)
+      .then(function (result) {
+        return _.extend(pluginArgs, result);
```
Can you please add a warning to the compilation if the result is undefined?

"`Using webpack-html-plugin without returning a result is deprecated.`"

jantimon replied:
Maybe I am just confused but the error occurs only if there are two plugins and the first one doesn't return a result right?

So if only one plugin is used which returns no value this should show the deprecated warning but still everything would work.

jantimon commented:
@ascariandrea I think that's a good idea - could you please add the (hopefull last two changes) ?

ascariandrea commented:
@jantimon I've added the warning and fixed the tests!
